internal/controller/controllers/agent_controller.go

```diff
@@ -77,6 +77,37 @@ type AgentReconciler struct {
 	ApproveCsrsRequeueDuration time.Duration
 }
 
+func (r *AgentReconciler) getClusterDeployment(agent *aiv1beta1.Agent, context *context.Context) (*hivev1.ClusterDeployment, error) {
+	kubeKey := types.NamespacedName{
+		Namespace: agent.Spec.ClusterDeploymentName.Namespace,
+		Name:      agent.Spec.ClusterDeploymentName.Name,
+	}
+	if agent.Spec.ClusterDeploymentName != nil {
+		clusterDeployment := &hivev1.ClusterDeployment{}
+		err := r.Get(*context, kubeKey, clusterDeployment)
+		if err == nil {
+			return clusterDeployment, nil
+		}
+		return nil, err
+	}
+	return nil, nil
+}
+
+func (r *AgentReconciler) getCluster(agent *aiv1beta1.Agent) (*common.Cluster, error){
+	kubeKey := types.NamespacedName{
+		Namespace: agent.Spec.ClusterDeploymentName.Namespace,
```

Comment:
You shouldn't access agent.Spec.ClusterDeploymentName before you know it's not nil